include/cocaine/detail/service/locator.hpp

```diff
@@ -114,8 +114,9 @@ class locator_t:
     // Outgoing router streams indexed by some arbitrary router-provided uuid.
     synchronized<router_map_t> m_routers;
 
-    std::uint32_t link_attempts;
-    synchronized<std::unique_ptr<asio::deadline_timer>> link_timer;
+    // This is here to store connection retry timers and to cancel them all on shutdown
+    // Map is used here to mimic C++17 behaviour of set.erase - ugly, isn't it?
+    std::map<asio::deadline_timer*, std::shared_ptr<asio::deadline_timer>> m_retry_timers;
```

Comment:
What's wrong with you, rebyatky? Use std::string (UUID) instead as a key.

Reply:
Also this sucks, because of thread safety. The destructor can run while there is `retry_link_node` doing its stuff.